api/middleware/responseLogger.go

```diff
@@ -0,0 +1,104 @@
+package middleware
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/gin-gonic/gin"
+)
+
+const prefixDurationTooLong = "[too long]"
+const prefixBadRequest = "[bad request]"
+const prefixInternalError = "[internal error]"
+
+type responseLoggerMiddleware struct {
+	thresholdDurationForLoggingRequest time.Duration
+}
+
+// NewResponseLoggerMiddleware returns a new instance of responseLoggerMiddleware
+func NewResponseLoggerMiddleware(thresholdDurationForLoggingRequest time.Duration) *responseLoggerMiddleware {
+	return &responseLoggerMiddleware{
+		thresholdDurationForLoggingRequest: thresholdDurationForLoggingRequest,
+	}
+}
+
+// MonitoringMiddleware logs detail about a request if it is not successful or it's duration is higher than a threshold
+func (rl *responseLoggerMiddleware) MiddlewareHandlerFunc() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		t := time.Now()
+
+		bw := &bodyWriter{body: bytes.NewBufferString(""), ResponseWriter: c.Writer}
+		c.Writer = bw
+
+		c.Next()
+
+		latency := time.Since(t)
+		status := c.Writer.Status()
+		if latency > rl.thresholdDurationForLoggingRequest || c.Writer.Status() != http.StatusOK {
+			responseBodyString := removeWhitespacesFromString(bw.body.String())
+			logRequestAndResponse(c, latency, status, responseBodyString)
+		}
+	}
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (rl *responseLoggerMiddleware) IsInterfaceNil() bool {
+	return rl == nil
+}
+
+func logRequestAndResponse(c *gin.Context, duration time.Duration, status int, response string) {
+	request := "n/a"
```

Comment:
As discussed, we can have a flag for this (defined in api.toml) to enable the log only when desired.

Reply:
done. added a flag for the `enabled` switch and also moved the threshold to the config file